proposals/lip-0049.md

```diff
@@ -223,168 +199,177 @@ When processing a cross-chain message `ccm`, the functions below may be called b
+            terminateChain(ccm.sendingChainID)
+            bounce(ccm, ERROR_CODE_INVALID_CCM)  
 ```
 
 #### forward
 
-The `forward` function forwards a CCM to the specified receiving chain. This function is only called on the mainchain when processing a CCU.
+The `forward` function forwards a CCM to the specified receiving chain. This function is only called on the mainchain when processing a CCU. If an error occurs, the CCM is bounced back to the sending chain or discarded.
 
 ```python
-# The ccu including this message is passed as context to this function
-forward(ccu, ccm):
-    if (chainAccount(ccm.receivingChainID) exists
-        and chainAccount(ccm.receivingChainID).status == CHAIN_ACTIVE
-        and isLive(ccm.receivingChainID):
-        # The Token module is called to transfer the fee from sending to receiving chain
-        token.forwardMessageFee(ccm)
-        if the previous call fails or returns False:
+def forward(ccu: CCU, ccm: CCM) -> None:
+    # Calculate CCM ID, used later in events.
+    ccmID = sha256(encode(crossChainMessageSchema, ccm))
+    
+    # Chain with status REGISTERED do not pass this check.
+    if isLive(ccm.receivingChainID) and chainAccount(ccm.receivingChainID).status == CHAIN_STATUS_ACTIVE:
+        
+        # Create a state snapshot.
+        baseSnapshot = snapshot of the current state
+        try:  
+            # Call the beforeCrossChainCommandForwarding functions from other modules
+            # For example, the Token module transfers the fee from escrow to escrow.
+            for each module mdl for which beforeCrossChainCommandForwarding exists:
+                mdl.beforeCrossChainCommandForwarding(ccu, ccm)
+        except:
+            revert state to baseSnapshot
+            terminateChain(ccm.sendingChainID)
+            bounce(ccm, ERROR_CODE_INVALID_CCM)
             return
-        addToOutbox(ccm.receivingChainID, ccm)
-        return
 
-    if ccm.status != CCM_STATUS_OK:
-        # Stop the processing of this message
+        addToOutbox(ccm.receivingChainID, ccm)
+        # Emit CCM forwarded event.
+        emitEvent(
+            module = MODULE_NAME_INTEROPERABILITY,
+            name = EVENT_NAME_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "result": CCM_FORWARDED},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
         return
 
-    # In all following cases, bounce the message with updated status
-    # Note: the fee was not assigned to the mainchain relayer and is also bounced
-    bounce(ccm)
+    bounce(ccm, ERROR_CODE_CHANNEL_UNAVAILABLE, ERROR_CODE_CHANNEL_UNAVAILABLE)
 
-    # If the chain does not exist or is not yet active don't continue
-    if (chainAccount(ccm.receivingChainID) does not exist
-        or chainAccount(ccm.receivingChainID).status == CHAIN_REGISTERED):  
+    # If the chain account does not exist, do not continue
+    if chainAccount(ccm.receivingChainID) does not exist:
+        return
```

Comment:
If I bounce here, I could provide more fine-grained error codes, i.e., channel does not exist, channel not activated below and finally channel terminated.

Reply:
For the moment I left it generic. If you guys think this is important, we can easily add more error codes

Reply:
I would find this would great some benefits and, considering it is a simple change now, I would do it.

Reply:
I added the fine-grained error codes and therefore resolve this comment.